### PR TITLE
Support for extra key-value fields. Edit images metadata.

### PR DESCRIPTION
**Pull request by burnash**
This is a first part of #234 <sup>[(see discussion)](https://github.com/NeuroVault/NeuroVault/pull/234#issuecomment-125796386)</sup>. I stripped the import-related features and refined the UI.

In particular: 

* Put back the "Add image" button.
* Marked required fixed fields the same way as on the Edit Image page.
* Made strict autocomplete with cell validation & highlighting.
* Added option to delete column of arbitrary metadata.
* Reworked metadata column addition.

### CODE REVIEW

**Comment by burnash** [comment]
@vsoch @chrisfilo Tweaked handsontable & submit button a bit, but overall it's ready for a review.

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+  function getNewColumnName(callback) {
+    var $modalEl = $('#columnNameModal'),
+      $submit = $('button[type="submit"]', $modalEl),
+      $columnNameEl = $('#column-name'),
+      $inputWrapper = $('.js-input-wrapper', $modalEl);
+
+    $modalEl.on('shown.bs.modal', function () {
+      $columnNameEl.focus().attr("autocomplete", "off");
+
+      $columnNameEl.keyup(function (e) {
```
Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Codacy Issue found: ['e' is defined but never used.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/825092758/issues/source?bid=2767031&fileBranchId=2767194#l217)

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js [hunk 2]
```diff
@@ -0,0 +1,382 @@
+/*global window, document, jQuery, Handsontable, NVMetadata */
+(function ($) {
+  'use strict';
+
+  function isString(val) {
```
Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Codacy Issue found: ['isString' is defined but never used.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/825092758/issues/source?bid=2767031&fileBranchId=2767194#l5)

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+        cells: function (r, c, prop) {
+          if (c === 0) {
+            return {
+              readOnly: true
+            };
+          }
+        }
+      });
+
+    $('.btn-save-metadata').click(function (e) {
```
Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Codacy Issue found: ['e' is defined but never used.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/825092758/issues/source?bid=2767031&fileBranchId=2767194#l350)

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+  }
+
+  function debounce(func, wait, immediate) {
+    var timeout;
+    return function () {
+      var context = this,
+        args = arguments;
+      var later = function () {
+        timeout = null;
+        if (!immediate) func.apply(context, args);
```
Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Codacy Issue found: [Expected '{' and instead saw 'func'.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/825092758/issues/source?bid=2767031&fileBranchId=2767194#l28)

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+              disabled: function () {
+                return isFixedColumn(hot.getSelected()[1])
+              }
+            },
+            'hsep3': '---------',
+            'undo': {},
+            'redo': {}
+          }
+        },
+        cells: function (r, c, prop) {
```
Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Codacy Issue found: ['prop' is defined but never used.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/825092758/issues/source?bid=2767031&fileBranchId=2767194#l341)

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+      var context = this,
+        args = arguments;
+      var later = function () {
+        timeout = null;
+        if (!immediate) func.apply(context, args);
+      };
+      var callNow = immediate && !timeout;
+      clearTimeout(timeout);
+      timeout = setTimeout(later, wait);
+      if (callNow) func.apply(context, args);
```
Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Codacy Issue found: [Expected '{' and instead saw 'func'.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/825092758/issues/source?bid=2767031&fileBranchId=2767194#l33)

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js [hunk 7]
```diff
@@ -0,0 +1,382 @@
+/*global window, document, jQuery, Handsontable, NVMetadata */
+(function ($) {
+  'use strict';
+
+  function isString(val) {
+    return (typeof val === 'string' || val instanceof String);
+  }
+
+  function cache() {
```
Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Codacy Issue found: [Function 'cache' has the same name as a variable declared in line 57.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/825092758/issues/source?bid=2767031&fileBranchId=2767194#l9)

File: neurovault/apps/statmaps/views.py [hunk 8]
```diff
@@ -174,6 +180,71 @@ def edit_collection(request, cid=None):
...
+
+    field_choices = ((f, statmap_field_obj(f).choices) for f in fixed_fields)
+
+    fields_with_choices = (t for t in field_choices if t[1])
+
+    return dict((field_name, pick_second_item(choices))
+                for field_name, choices in fields_with_choices)
+
+
+def cognitive_atlas_task_datasource(request):
```
Comment (chrisfilo) [inline neurovault/apps/statmaps/views.py]: Codacy Issue found: [map/filter on lambda could be replaced by comprehension](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/825092872/issues/source?bid=2767031&fileBranchId=2767194#l199)

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js (hunk 1 above)
Comment (burnash) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Removed 'e' in 4bdc65d86

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js (hunk 2 above)
Comment (burnash) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Removed 'isString' in 4bdc65d86

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js (hunk 3 above)
Comment (burnash) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Removed 'e' in 4bdc65d86

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js (hunk 4 above)
Comment (burnash) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Added  '{' in 4bdc65d86

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js (hunk 6 above)
Comment (burnash) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Added '{' in 4bdc65d86

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js (hunk 5 above)
Comment (burnash) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Removed 'prop' in 4bdc65d86

File: neurovault/apps/statmaps/static/scripts/metadata-edit.js (hunk 7 above)
Comment (burnash) [inline neurovault/apps/statmaps/static/scripts/metadata-edit.js]: Renamed 'cache' in 4bdc65d86

File: neurovault/apps/statmaps/views.py (hunk 8 above)
Comment (burnash) [inline neurovault/apps/statmaps/views.py]: Replaced W0110 in 5d208c02c1

**Comment by chrisfilo** [comment]
Looks great! Couple of things:

* Could you write that columns with "*" are compulsory on the edit page?
* Could you use human readable column names (instead of field names)
* The error reporting is works but is not the prettiest. Would it be difficult to implement faulty cell highlighting like @vsoch suggested in https://github.com/NeuroVault/NeuroVault/pull/234#issuecomment-125691403

This feature is going to be amazing when we finally merge it! @GaelVaroquaux and @satra will definitely appreciate custom key-valye tags in bulk editing!

**Comment by vsoch** [comment]
+1 !

**Comment by burnash** [comment]
@chrisfilo thanks for the review.

> * Could you write that columns with "*" are compulsory on the edit page?

Sure, I made it the same way as on the [Edit Image page](https://github.com/NeuroVault/NeuroVault/blob/0019052757d70b8e71eb388116ea86373f8faad7/neurovault/apps/statmaps/templates/statmaps/edit_image.html.haml), where this message is missing. I thought it's intentionally missing. To save some time on "trial and error" page design cycle and related discussions, could you specify how exactly this message should look like or reference some page element somewhere on NeuroVault I can borrow a style from?
<br/>

>  * Could you use human readable column names (instead of field names)

I can, but wouldn't it create inconsistency and confusion for the user? On the details tab of an image page these fields are listed with field names (as in the metadata grid):

screen shot 2015-08-03 at 18 31 51
<br/>

> * The error reporting is works but is not the prettiest. Would it be difficult to implement faulty cell highlighting like @vsoch suggested in #234 (comment)

Could you elaborate on this? With test cases and screenshots for clarity if possible. The error reporting is made exactly as in @vsoch's comment and does have a faulty cell highlighting.

Missing value cell highlight (before submit):

screen shot 2015-08-03 at 18 55 06

Incorrect & missing value cells highlight along with a server validation error message:

screen shot 2015-08-03 at 18 59 43

**Comment by chrisfilo** [comment]
I'm having a retarded morning and edited your comment instead of replying to it! Sorry!


>> * Could you write that columns with "*" are compulsory on the edit page?

>Sure, I made it the same way as on the [Edit Image page](https://github.com/NeuroVault/NeuroVault/blob/0019052757d70b8e71eb388116ea86373f8faad7/neurovault/apps/statmaps/templates/statmaps/edit_image.html.haml), where this message is missing. I thought it's intentionally missing. To save some time on "trial and error" page design cycle and related discussions, could you specify how exactly this message should look like or reference some page element somewhere on NeuroVault I can borrow a style from?
<br/>

Ok let's leave this for now.

>>  * Could you use human readable column names (instead of field names)

>I can, but wouldn't it create inconsistency and confusion for the user? On the details tab of an image page these fields are listed with field names (as in the metadata grid):
screen shot 2015-08-03 at 18 31 51
<br/>

You are right. I will correct the "details" tab in a separate PR to make it consistent. The Edit and add image forms use the human readable names thought.

>> * The error reporting is works but is not the prettiest. Would it be difficult to implement faulty cell highlighting like @vsoch suggested in #234 (comment)

>Could you elaborate on this? With test cases and screenshots for clarity if possible. The error reporting is made exactly as in @vsoch's comment and does have a faulty cell highlighting.
Missing value cell highlight (before submit):
screen shot 2015-08-03 at 18 55 06
Incorrect & missing value cells highlight along with a server validation error message:
screen shot 2015-08-03 at 18 59 43

Yeah this did not work for me. Steps to replicate:
1. clean the environment "docker-compose rm" and rebuild the neurovault docker image
2. docker-compose up, go to website create a user
3. create a collection and add a bunch of maps using the upload folder interface
4. Edit images metadata
5. Since you uploaded via folder upload many filed will have missing values
6. Click save
7. No cells are highlighted. Errors concerning only one column are displayed.

screen shot 2015-08-03 at 09 50 46



JS consol reports this error:

	POST http://192.168.59.103/collections/1/editmetadata 400 (BAD REQUEST)send @ jquery-1.10.2.min.js:6x.extend.ajax @ jquery-1.10.2.min.js:6(anonymous function) @ metadata-edit.js:354x.event.dispatch @ jquery-1.10.2.min.js:5v.handle @ jquery-1.10.2.min.js:5

**Comment by burnash** [comment]
@chrisfilo Thanks for the detailed description of the "highlighting" test case. That was due to the way NeuroVault currently allows missing values in required fields when mass-uploading images. I've figured out how to pass cogatlas paradigm values to django's model validation handlers and now it highlights faulty cells properly.

> You are right. I will correct the "details" tab in a separate PR to make it consistent. The Edit and add image forms use the human readable names thought.

I've added human readable (aka *verbose names*) to the grid header in 3d0e971.

Now the grid in your test case looks like this:

screen shot 2015-08-04 at 19 38 59

> JS consol reports this error:
> POST http://192.168.59.103/collections/1/editmetadata 400 (BAD REQUEST)

It's not an error per se. When ajax validation fails I return a json with errors and a HTTP code 400 (bad request). There's a [long](http://stackoverflow.com/questions/3290182/rest-http-status-codes-for-failed-validation-or-invalid-duplicate) [running](http://stackoverflow.com/questions/1959947/whats-an-appropriate-http-status-code-to-return-by-a-rest-api-service-for-a-val) [debate](http://stackoverflow.com/questions/362618/proper-use-of-http-status-codes-in-a-validation-server) on what is a proper status. I stick with 400 since it's [the same code as in the Django REST framework](http://www.django-rest-framework.org/api-guide/exceptions/) and you're already using it.

**Comment by chrisfilo** [comment]
Excellent!

There is one more tiny thing I noticed in your screenshot: "Brain coverage".
We have a set of fields that are not exposed through the form such as: "brain_coverage", "perc_voxels_outside", "is_thresholded" etc. Those are calculated automatically and should not be editable.

Could you restrict the columns of the table to fields that are editable? This might be a little bit tricky. We basically need fields listed here: https://github.com/NeuroVault/NeuroVault/blob/master/neurovault/apps/statmaps/forms.py#L553
with the exception of those that have hidden input widget https://github.com/NeuroVault/NeuroVault/blob/master/neurovault/apps/statmaps/forms.py#L491

Maybe there is a way to take AddStatisticMapForm list all fields and corresponding widgets and include only the fields that do not use "HiddenInput" widget. See https://docs.djangoproject.com/en/1.8/ref/forms/api/#accessing-the-fields-from-the-form and https://docs.djangoproject.com/en/1.8/ref/forms/fields/#widget. I'm also open to other solutions.

On that note I also expect that the tabular edit interface will brake for heterogenous collections (collections that include not only statistical maps but also atlases and NIDM files). For such collections I would just turn off this option for now.

**Comment by burnash** [comment]
@chrisfilo

> Could you restrict the columns of the table to fields that are editable? This might be a little bit tricky.
> [...]  I'm also open to other solutions.

An alternative, somewhat simpler and robust solution is already implemented. I was looking for a way to define "fixed" fields which should go to the tabular interface. Since different fields are defined at different levels of model inheritance hierarchy I figured out explicit way to define them. At each level a model has a [`get_fixed_fields`](https://github.com/NeuroVault/NeuroVault/blob/8c2351cdf837c75d4e80a3094407244a2e104705/neurovault/apps/statmaps/models.py#L470) which returns a tuple with "fixed" fields plus parent class's fixed fields:

~~~python
@classmethod
def get_fixed_fields(cls):
    return super(BaseStatisticMap, cls).get_fixed_fields() + (
        'map_type', 'analysis_level')
~~~

I removed `brain_coverage` in 8c2351c. Now there are only editable fields in the tabular interface.

Is there any reason not to use [`editable=False`](https://docs.djangoproject.com/en/1.7/ref/models/fields/#editable) attribute for those hidden model fields?

> On that note I also expect that the tabular edit interface will brake for heterogenous collections (collections that include not only statistical maps but also atlases and NIDM files). For such collections I would just turn off this option for now.

Yes, we can turn off a menu item on the details page and the tabular page itself if a collection contains something other than stat maps.

Is there any efficient way to find out if a collection contains only stat maps? One way I can think of is to define a property on `Collection` model:

~~~python
class Collection(models.Model):
    ...
    @property
    def statistic_maps_only(self):
        return all((isinstance(i, StatisticMap) for i in self.image_set.all()))
~~~

(There must be a better name for it, though.)

Then use this property in templates & views.

**Comment by chrisfilo** [comment]
Oh did not know about "editable" - if this indeed works and we switch to that would it make the "get_fixed_fields" method redundant?

I would suggest renaming the Collection property to "statisticmap_set" to be in line with "image_set".

I tested the validation code - it works correctly!

**Comment by burnash** [comment]
> Oh did not know about "editable" - if this indeed works and we switch to that would it make the "get_fixed_fields" method redundant?

Yes, if there're no other restrictions on fixed fields displayed in the tabular interface. In other words, will we output all editable fields in the tabular interface? If yes, I propose to open up a separate issue to implement `editable` attribute usage.

> I would suggest renaming the Collection property to "statisticmap_set" to be in line with "image_set".

This is a better name, I agree. However since this method returns boolean value, I suggest to add a prefix "is_" so it'll read `is_statisticmap_set`. What do you think?

**Comment by chrisfilo** [comment]
On Aug 6, 2015 7:57 AM, "Anton Burnashev" <notifications@github.com> wrote:
>>
>> Oh did not know about "editable" - if this indeed works and we switch to
that would it make the "get_fixed_fields" method redundant?
>
> Yes, if there're no other restrictions on fixed fields displayed in the
tabular interface. In other words, will we output all editable fields in
the tabular interface? If yes, I propose to open up a separate issue to
implement editable attribute usage.
I'll do that shortly, but let's not wait until that's done with this PR.
>>
>> I would suggest renaming the Collection property to "statisticmap_set"
to be in line with "image_set".
>
> This is a better name, I agree. However since this method returns boolean
value, I suggest to add a prefix "is_" so it'll read is_statisticmap_set.
What do you think?
+1

**Comment by burnash** [comment]
@chrisfilo Great! I've added the property and blocked access to the edit metadata page in a97d475
Next push will rebase the branch.

File: neurovault/apps/statmaps/static/scripts/papaparse.min.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
...
...tElem);if(typeof n==="object"){if(n.action=="abort"){s("AbortError",e.file,e.inputElem,n.reason);return}else if(n.action=="skip"){u();return}else if(typeof n.config==="object")e.instanceConfig=o.extend(e.instanceConfig,n.config)}else if(n=="skip"){u();return}}var i=e.instanceConfig.complete;e.instanceConfig.complete=function(t){if(x(i))i(t,e.file,e.inputElem);u()};Papa.parse(e.file,e.instanceConfig)}function s(e,n,r,i){if(x(t.error))t.error({name:e},n,r,i)}function u(){r.splice(0,1);i()}var n=t.config||{};var r=[];this.each(function(t){var i=o(this).prop("tagName").toUpperCase()=="INPUT"&&o(this).attr("type").toLowerCase()=="file"&&e.FileReader;if(!i||!this.files||this.files.length==0)return true;for(var s=0;s<this.files.length;s++){r.push({file:this.files[s],inputElem:this,instanceConfig:o.extend({},n)})}});i();return this}}if(t){e.onmessage=w}else if(Papa.WORKERS_SUPPORTED){r=v();if(!document.body){n=true}else{document.addEventListener("DOMContentLoaded",function(){n=true},true)}}l.prototype=Object.create(f.prototype);l.prototype.constructor=l;c.prototype=Object.create(f.prototype);c.prototype.constructor=c;h.prototype=Object.create(h.prototype);h.prototype.constructor=h})(this);
```
Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/papaparse.min.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (13)](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/836105118/issues/source?bid=2767031&fileBranchId=2767194#l6)

Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/papaparse.min.js]: Codacy Issue found: [Expected '===' and instead saw '=='.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/836105118/issues/source?bid=2767031&fileBranchId=2767194#l6)

Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/papaparse.min.js]: Codacy Issue found: [The body of a for in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/836105118/issues/source?bid=2767031&fileBranchId=2767194#l6)

Comment (chrisfilo) [inline neurovault/apps/statmaps/static/scripts/papaparse.min.js]: Codacy Issue found: ['r' is defined but never used.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/836105118/issues/source?bid=2767031&fileBranchId=2767194#l6)

**Comment by burnash** [comment]
@chrisfilo Rebased and removed some unused js dependencies remained from the import feature.

**Comment by burnash** [comment]
@chrisfilo @vsoch Thank you guys for your help with this PR and for getting it live!

File: neurovault/apps/statmaps/migrations/0038_image_data.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    dependencies = [
+        ('statmaps', '0037_auto_20150307_0222'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='image',
+            name='data',
+            field=django_hstore.fields.DictionaryField(null=True, blank=True),
+            preserve_default=True,
```
Comment (chrisfilo) [inline neurovault/apps/statmaps/migrations/0038_image_data.py]: This should have a default = {}. We need better tests for deploying new versions on old data.

Comment (burnash) [inline neurovault/apps/statmaps/migrations/0038_image_data.py]: I agree. How the problem manifested itself?

Comment (chrisfilo) [inline neurovault/apps/statmaps/migrations/0038_image_data.py]: API calls were failing for all images added before merging this PR. I had
to set 'data' to {} for all images.

**Comment by chrisfilo** [comment]
Excellent work @burnash!

**Pull request merged**